google-cloud-firestore/src/main/java/com/google/cloud/firestore/BulkWriter.java

```diff
@@ -0,0 +1,763 @@
+   * @param precondition Precondition to enforce on this update.
+   * @param fieldPath The first field to set.
+   * @param value The first value to set.
+   * @param moreFieldsAndValues String and Object pairs with more fields to be set.
+   * @return An ApiFuture containing the result of the write. Contains an error if the write fails.
+   */
+  @Nonnull
+  public ApiFuture<WriteResult> update(
+      @Nonnull DocumentReference documentReference,
+      @Nonnull Precondition precondition,
+      @Nonnull FieldPath fieldPath,
+      @Nullable Object value,
+      Object... moreFieldsAndValues) {
+    verifyNotClosed();
+    BulkCommitBatch bulkCommitBatch = getEligibleBatch(documentReference);
+    ApiFuture<WriteResult> future =
+        bulkCommitBatch.update(
+            documentReference, precondition, fieldPath, value, moreFieldsAndValues);
+    sendReadyBatches();
+    return future;
+  }
+
+  /**
+   * Commits all writes that have been enqueued up to this point in parallel.
+   *
+   * <p>Returns an ApiFuture that completes when all currently queued operations have been
+   * committed. The ApiFuture will never return an error since the results for each individual
+   * operation are conveyed via their individual ApiFutures.
+   *
+   * <p>The ApiFuture completes immediately if there are no pending writes. Otherwise, the ApiFuture
+   * waits for all previously issued writes, but it does not wait for writes that were added after
+   * the method is called. If you want to wait for additional writes, call `flush()` again.
+   *
+   * @return An ApiFuture that completes when all enqueued writes up to this point have been
+   *     committed.
+   */
+  @Nonnull
+  public ApiFuture<Void> flush() {
+    verifyNotClosed();
+    final SettableApiFuture<Void> flushComplete = SettableApiFuture.create();
+    List<SettableApiFuture<WriteResult>> writeFutures = new ArrayList<>();
+    for (BulkCommitBatch batch : batchQueue) {
+      batch.markReadyToSend();
+      writeFutures.addAll(batch.getPendingFutures());
+    }
+    sendReadyBatches();
+    ApiFutures.successfulAsList(writeFutures)
+        .addListener(
+            new Runnable() {
+              public void run() {
+                flushComplete.set(null);
+              }
+            },
+            MoreExecutors.directExecutor());
+    return flushComplete;
+  }
+
+  /**
+   * Commits all enqueued writes and marks the BulkWriter instance as closed.
+   *
+   * <p>After calling `close()`, calling any method wil return an error.
+   *
+   * <p>Returns an ApiFuture that completes when there are no more pending writes. The ApiFuture
+   * will never error. Calling this method will send all requests. The ApiFuture completes
+   * immediately if there are no pending writes.
+   *
+   * @return An ApiFuture that completes when all enqueued writes up to this point have been
+   *     committed.
+   */
+  @Nonnull
+  public ApiFuture<Void> close() {
```

Comment:
It appears this method is only used in tests right now. Can we scope it to package private, or is the intent that users will call this themselves?

There are a couple of interfaces in java, notably the general purpose [`java.lang.AutoClosable`](https://docs.oracle.com/javase/8/docs/api/index.html?java/lang/AutoCloseable.html#close--) and the IO specific [`java.io.Closable`](https://docs.oracle.com/javase/8/docs/api/java/io/Closeable.html#close--) which already have defined method signatures for `close`, specifically `void close() ...` if we ship this public signature then we would prevent the ability to use a BulkWriter in a [try-with-resources](https://docs.oracle.com/javase/tutorial/essential/exceptions/tryResourceClose.html) context.

If we do want to keep the method public, can we rename it to something like `completeAll` or `closeAll` to remove the name collision?

Reply:
Follow up from conversation:
"AutoClosable already dictates the method signature of close as `public void close() throws Exception` so, the existing `ApiFuture<Void> close()` will collide with it and result in a compilation error."

Solution: Change `close()`to synchronous method that blocks on the writes, which will allow the override to work and still support AutoClosable.